docs/build/advanced-concepts/randomness.md

```diff
@@ -0,0 +1,177 @@
+- If the outcome is unfavorable (based on some criteria codified in the transaction), the transaction aborts itself.
+
+As an example, consider a simple coin toss randomized contract where users can bet any amount of tokens against a random binary output. If the coin toss contract outputs `1`, the user doubles their bet. If the coin toss contract outputs `0`, the user loses their bet in favor of the coin toss.
+
+Although the user (or the honest coin toss contract) cannot predict or bias the outcome, the user transaction can check the randomized result and cancel the transaction if they are losing their bet. This can be done by calling an exception causing the transaction to error. All temporary state changes are cancelled and the user can repeat the process till they double their bet.
+
+## Commit-Reveal Scheme
+
+The recommended way to mitigate the problems above is via a commit-reveal scheme. The scheme involves two steps: commit and reveal. During the commit phase, users submit a commitment that contains the hash of their answer along with a random seed value. The smart contract stores this commitment on the blockchain. Later, during the reveal phase, the user reveals their answer and the seed value.
+
+There are ideas how to further optimize the developer experience further in the future. For example, a transaction could delegate part of its gas to an independent transaction it spawns. Conceptually, also this future solution would be a commit-and-reveal scheme, just immediately happening within the same block. Until we eventually get to this next level, developers may need to implement their own commit-reveal. In Cadence, it is clean and short.
+
+Protocol improvements (documented in [FLIP 120](https://github.com/onflow/flips/blob/main/cadence/20230713-random-function.md)) expose the random beacon to the FVM and Cadence where it can be used to seed pseudo-random number generators (PRG) for smart contracts.
+
+In simpler terms, the native secure randomness provided by the protocol can now be safely utilized within Cadence smart contracts and is available to all developers on Flow and the FVM.
+
+While the source values are safely generated by the Random Beacon and transmitted into the execution environment via the committing transaction, using the raw values from this contract does not guarantee non-revertible randomness. In addition to the commit-reveal pattern described below, it is also recommended to use the source values with a pseudo-random number generator (PRG) to generate an arbitrarily-long sequence of random values.
+
+Adding a safe pattern to reveal randomness without the possibility of conditional transaction reversion unlocks applications relying on randomness. By providing examples of commit-reveal implementations we hope to foster a more secure ecosystem of decentralized applications and encourage developers to build with best practices.
```

Comment:
```suggestion
The recommended way to mitigate the problems above is via a commit-reveal scheme.
The scheme involves two steps: commit and reveal. During the commit phase, the user
transaction commits to accepting the future output of a smart contract where the last
remaining input is an unknown stream of random numbers. The smart contract stores this
commitment on the blockchain. At the current level of optimization, the reveal phase can
start as early as the next block, when the beacon's source of randomness becomes available.
With a second transaction, the smart contract can be executed to explicitly generate the
output now that the stream of random numbers can be seeded (plus suitable salt).

There are ideas how to optimize the developer experience further in the future.
For example, a transaction could delegate part of its gas to an independent transaction
it spawns. Such solution would conceptually also be a commit-and-reveal scheme, just immediately
happening within the same block. Until we eventually get to this next level, developers may need
to implement their own commit-reveal. In Cadence, it is clean and short.

Protocol improvements (documented in [FLIP 120](https://github.com/onflow/flips/blob/main/cadence/20230713-random-function.md))
expose the random beacon to the FVM and Cadence where it can be used to seed pseudo-random number generators [PRNGs] for smart contracts.

While both are backed by Flow's Random Beacon it is important for developers to mindfully choose between `revertibleRandom` or
seeding their own PRNG utilizing the `RandomBeaconHistory` smart contract:

- Under the hood, the FVM also just instantiates a PRNG for each transaction that `revertibleRandom` draws from.
  Though, with `revertibleRandom` a developer is calling the PRNG that is controlled by the transaction,
  which also has the power to abort and revert if it doesn't like `revertibleRandom`'s outputs.
  `revertibleRandom` is only suitable for smart contract functions that exclusively run within the trusted transactions.
- In contrast, using the `RandomBeaconHistory` means to use a deterministically-seeded PRNG.
  The `RandomBeaconHistory` is key for effectively implementing a commit-and-reveal scheme.
  During the commit phase, the user commits to proceed with a future source of randomness,
  which is revealed after the commit transaction concluded.
  For each block, the `RandomBeaconHistory` automatically stores the subsequently generated source of randomness.

Adding a safe pattern to reveal randomness without the possibility of conditional transaction reversion unlocks applications relying on randomness.
By providing examples of commit-reveal implementations we hope to foster a more secure ecosystem of decentralized applications
and encourage developers to build with best practices.

In simpler terms, the native secure randomness provided by the protocol can now be safely utilized within Cadence smart contracts
and is available to all developers on Flow and the FVM.
```